src/client.py

```diff
@@ -1,45 +1,83 @@
 from . import errors, defines
 from .block import Block
 from .connection import Connection
+from .progress import Progress
 from .protocol import ServerPacketTypes
 from .util.escape import escape_params
 from .util.helpers import chunks
 
 
 class QueryResult(object):
-    def __init__(self, with_column_types=False):
+    def __init__(
+            self, packet_generator,
+            with_column_types=False, columnar=False):
+        self.packet_generator = packet_generator
         self.with_column_types = with_column_types
 
         self.data = []
         self.columns_with_types = []
+        self.columnar = columnar
 
         super(QueryResult, self).__init__()
 
+    def store(self, packet):
+        block = getattr(packet, 'block', None)
+        if block is None:
+            return
+
+        # Header block contains no rows. Pick columns from it.
+        if block.rows:
+            if self.columnar:
+                columns = block.get_columns()
+                if self.data:
+                    # Extend corresponding column.
+                    for i, column in enumerate(columns):
+                        self.data[i] += column
+                else:
+                    self.data.extend(columns)
+            else:
+                self.data.extend(block.get_rows())
+
+        elif not self.columns_with_types:
+            self.columns_with_types = block.columns_with_types
+
     def get_result(self):
+        for packet in self.packet_generator:
+            self.store(packet)
+
         if self.with_column_types:
             return self.data, self.columns_with_types
         else:
             return self.data
 
 
-class Progress(object):
-    def __init__(self, client_connection, result, progress_gen):
-        self.connection = client_connection
-        self.result = result
-        self.progress_gen = progress_gen
+class ProgressQueryResult(QueryResult):
```

Comment:
Let's refactor a little bit and place `ProgressQueryResult` and `QueryResult` in `result.py` file.

Reply:
Good idea, done